### PR TITLE
fix unnecessary recursion

### LINKED ISSUE #166: unnecessary recursion
[bracket-matcher.coffee#L86](https://github.com/atom/bracket-matcher/blob/21ac27bc870780be08459b8cc7287803639966e0/lib/bracket-matcher.coffee#L84) should probably be calling the original `@editor.insertText`, currently it calls the one created by `_.advise` on [L36](https://github.com/atom/bracket-matcher/blob/21ac27bc870780be08459b8cc7287803639966e0/lib/bracket-matcher.coffee#L36) which means bracket matcher needlessly gets to evaluate its own insertion for further matching.

### PR DESCRIPTION
**Pull request by jacekkopecky**
fixes #166

### CODE REVIEW

**Comment by thomasjo** [comment]
On first glance, this looks good to me. Hopefully build is :green_heart:

File: lib/bracket-matcher.coffee [hunk 1]
```diff
@@ -81,7 +82,7 @@ class BracketMatcher
       @editor.moveRight()
       false
     else if autoCompleteOpeningBracket
-      @editor.insertText(text + pair)
+      @origEditorInsertText(text + pair)
       @editor.moveLeft()
```
Comment (jeancroy) [inline lib/bracket-matcher.coffee]: On line 86. 

Before PR, the editor that `.moveLeft()` is the same as the editor that `.insertText()`.

Now they can potentially be two different editor. (If they are garantee to alwais the same I do not see the point of the PR). I'm not sure why a special treatment to the insertText method. There's a bunch of stuff done with the @editor object. Maybe it's that pointer that need to be protected against "unnecessary recursion" ?

Comment (jacekkopecky) [inline lib/bracket-matcher.coffee]: Hi. `@origEditorInsertText` is the same as `@editor.insertText` was before we did `_.adviseBefore` on it. 

The recursion is that `BracketMatcher::insertText` calls `BracketMatcher::insertText` again when it calls `@editor.insertText`, then decides nontrivially that it doesn't have to do anything and returns `true` which causes `adviseBefore` to finally call the original `TextEditor::insertText`.

Comment (jeancroy) [inline lib/bracket-matcher.coffee]: Ok I've digg into adviceBefore, it's like an event system
https://github.com/atom/underscore-plus/blob/master/src/underscore-plus.coffee#L57-L61

wouldn't just not doing the advice / commenting [the line](https://github.com/jacekkopecky/bracket-matcher/blob/fix-166/lib/bracket-matcher.coffee#L37) do the same thing ?

Comment (jacekkopecky) [inline lib/bracket-matcher.coffee]: @jeancroy `adviceBefore` is how bracket matcher intercepts user input and reacts to opening brackets by automatically inserting the closing ones. Commenting that line out would disable bracket matcher.

**Comment by smashwilson** [comment]
👍 Excellent. Thanks!

**Comment by jacekkopecky** [comment]
Wow, surprised the PR still worked. 8-)

**Pull request merged**